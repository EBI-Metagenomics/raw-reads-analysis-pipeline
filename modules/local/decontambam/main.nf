process DECONTAMBAM {
    tag "${meta.id}"
    label 'process_low'

    // conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/biocontainers/samtools:1.21--h50ea8bc_0'
        : 'quay.io/biocontainers/samtools:1.21--h50ea8bc_0'}"

    input:
    tuple val(meta), path(inputbam), val(split), val(fn_prefix)

    output:
    tuple val(meta), path("*_mapped{_1,_2,_interleaved}.fq.gz")  , emit: mapped_reads
    tuple val(meta), path("*_unmapped{_1,_2,_interleaved}.fq.gz"), emit: unmapped_reads
    tuple val(meta), path("*_all_summary_stats.txt")             , emit: stats
    tuple val(meta), path("*_mapped_summary_stats.txt")          , emit: mapped_stats
    tuple val(meta), path("*_unmapped_summary_stats.txt")        , emit: unmapped_stats
    path "versions.yml"                                          , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args1 = task.ext.args1 ?: ''
    def args2 = task.ext.args2 ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def all_prefix = "${prefix}_${fn_prefix}_all"
    def mapped_prefix = "${prefix}_${fn_prefix}_mapped"
    def unmapped_prefix = "${prefix}_${fn_prefix}_unmapped"
    if (split) {
        """
        #!/bin/bash
        samtools view \\
            ${args1} \\
            -f 4 \\
            -b ${inputbam} \\
        | samtools bam2fq \\
            ${args2} \\
            -@ ${task.cpus} \\
            -1 ${unmapped_prefix}_1.fq.gz \\
            -2 ${unmapped_prefix}_2.fq.gz \\
            -0 ${unmapped_prefix}_other.fq.gz \\
            -s ${unmapped_prefix}_singleton.fq.gz \\
            -

        samtools view \\
            ${args1} \\
            -F 4 \\
            -b ${inputbam} \\
        | samtools bam2fq \\
            ${args2} \\
            -@ ${task.cpus} \\
            -1 ${mapped_prefix}_1.fq.gz \\
            -2 ${mapped_prefix}_2.fq.gz \\
            -0 ${mapped_prefix}_other.fq.gz \\
            -s ${mapped_prefix}_singleton.fq.gz \\
            -

        samtools stats ${inputbam} \\
        > ${all_prefix}_summary_stats.txt

        samtools stats -F 4 ${inputbam} \\
        > ${mapped_prefix}_summary_stats.txt

        samtools stats -f 4 ${inputbam} \\
        > ${unmapped_prefix}_summary_stats.txt

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
        END_VERSIONS
        """
    }
    else {
        """
        #!/bin/bash
        samtools view \\
            ${args1} \\
            -f 4 \\
            -b ${inputbam} \\
        | samtools bam2fq \\
            ${args2} \\
            -@ ${task.cpus} \\
            - \\
        | gzip --no-name > ${unmapped_prefix}_interleaved.fq.gz

        samtools view \\
            ${args1} \\
            -F 4 \\
            -b ${inputbam} \\
        | samtools bam2fq \\
            ${args2} \\
            -@ ${task.cpus} \\
            - \\
        | gzip --no-name > ${mapped_prefix}_interleaved.fq.gz

        samtools stats ${inputbam} \\
        > ${all_prefix}_summary_stats.txt

        samtools stats -F 4 ${inputbam} \\
        > ${mapped_prefix}_summary_stats.txt

        samtools stats -f 4 ${inputbam} \\
        > ${unmapped_prefix}_summary_stats.txt

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
        END_VERSIONS
        """
    }
}
