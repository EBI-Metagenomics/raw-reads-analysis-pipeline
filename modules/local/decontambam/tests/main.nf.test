nextflow_process {

    name "Test Process DECONTAMBAM"
    script "../main.nf"
    process "DECONTAMBAM"

    tag "modules"
    tag "modules_"
    tag "decontambam"

    test("mini bam should match snapshot") {

        when {
            process {
                """
                input[0] = [
		    ['id': 'test'],
		    file("${projectDir}/tests/fixtures/decontambam/ERR10889188_mini.bam", checkIfExists: true),
		    true,
		    'test_prefix'
		]
                """
            }
        }

        then {
            assert process.success
            assert snapshot(process.out).match()
        }

    }

}
