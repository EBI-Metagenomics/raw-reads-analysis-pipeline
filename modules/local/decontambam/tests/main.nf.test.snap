{
    "mini bam should match snapshot": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_test_prefix_mapped_1.fq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_test_prefix_mapped_2.fq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_test_prefix_unmapped_1.fq.gz:md5,43986effab9cde0f7a8b94b7f1129104",
                            "test_test_prefix_unmapped_2.fq.gz:md5,1ed5f437da92c48e29a5f6def30c57ed"
                        ]
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_test_prefix_all_summary_stats.txt:md5,afb7839147f6fd34068446fc4b46707d"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_test_prefix_mapped_summary_stats.txt:md5,811a771d4ed803ef1bc2aadeee4030b4"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test_test_prefix_unmapped_summary_stats.txt:md5,ac9d67b5a61d9424cb7838e124d8de9a"
                    ]
                ],
                "5": [
                    "versions.yml:md5,3aefe6330e07b09a8c472e5296ee11c8"
                ],
                "mapped_reads": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_test_prefix_mapped_1.fq.gz:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_test_prefix_mapped_2.fq.gz:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "mapped_stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test_test_prefix_mapped_summary_stats.txt:md5,811a771d4ed803ef1bc2aadeee4030b4"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test_test_prefix_all_summary_stats.txt:md5,afb7839147f6fd34068446fc4b46707d"
                    ]
                ],
                "unmapped_reads": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_test_prefix_unmapped_1.fq.gz:md5,43986effab9cde0f7a8b94b7f1129104",
                            "test_test_prefix_unmapped_2.fq.gz:md5,1ed5f437da92c48e29a5f6def30c57ed"
                        ]
                    ]
                ],
                "unmapped_stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test_test_prefix_unmapped_summary_stats.txt:md5,ac9d67b5a61d9424cb7838e124d8de9a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3aefe6330e07b09a8c472e5296ee11c8"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-05-20T13:00:01.591054112"
    }
}